### PR TITLE
fix: enforce per-session send buffer limits before queueing

### PR DESCRIPTION
**Pull request by Officeyutong**
## Summary

Fix a send-buffer limit bypass that allowed slow or non-reading peers to accumulate outbound data beyond the configured `send_buffer_size`.

## Changes

- Atomically reserve pending bytes before protocol messages enter the service queue.
- Reject oversized reservations and close the affected session.
- Carry reservation state through `ServiceTask` to prevent double accounting.
- Reconcile reserved bytes when `before_send` changes message size.
- Roll back reservations when message delivery fails or the session is closed.
- Release pending bytes when messages target unavailable protocols.
- Enforce the limit independently of session-buffer state for both TCP/Yamux and QUIC.
- Add tests for admission limits, overflow, rollback, and early protocol-message accounting.

## Security Impact

Previously, messages could leave the session-level buffer while remaining queued in lower layers. Because limit enforcement depended on the session buffer being non-empty, a slow peer could bypass the configured byte limit and cause excessive memory consumption.

Messages are now accounted for before queue admission, so an over-limit message is rejected before it can consume additional queued memory.

## Validation

- `cargo check -p tentacle --all-targets`
- `cargo test -p tentacle`
- All unit and integration tests passed.
- End-to-end reproduction with 128 distinct 1 MiB messages and a 24 MiB limit:
  - `SessionBlocked` triggered
  - pending bytes returned to zero
  - peak RSS reduced from approximately 137 MiB to 31 MiB

### CODE REVIEW

**Comment by chainTe** [comment]
## Security review

Result: **2 confirmed findings (1 Medium, 1 Low)**

Reviewed range: `284470ee2ad7d78f4eea37a4267a73141cfa3c97..ee31ff39ce08d300d56b4e5700e9fe5944becdb7`  
Review model: **GPT-5.6 Codex**

### Medium — public service-control sends are accounted too late

`ServiceControl::{filter_broadcast,quick_filter_broadcast}` and the asynchronous equivalents put the complete `Bytes` payload into the shared service-task queue with `reserved: None` (`tentacle/src/service/control.rs:128-155`, `392-423`). The per-session limit is first applied after dequeue in `InnerService::handle_message` (`tentacle/src/service.rs:805-876`).

This means `set_send_buffer_size` does not bound all queued outbound bytes as documented. Current CKB protocol callbacks clone these public controls, and an untrusted peer's GetBlocks request can schedule multiple block responses through `ServiceAsyncControl`, so retained bytes can exceed the default 24 MiB session budget before admission runs. The queue's entry cap and protocol frame caps bound the amplification, which is why I rate this Medium rather than High.

Recommendation: resolve concrete targets and acquire a per-session RAII byte reservation before the first queue retains the payload. Broadcasts need one reservation per target.

### Low — cancelling a pending send leaks its reservation

`ProtocolContextMutRef::send_message_inner` reserves bytes and then awaits service-queue admission, but rolls back only after the await returns an error (`tentacle/src/context.rs:524-543`). While the queue is full, cancelling that future drops the payload before a downstream guard exists, leaving `pending_data_size` charged permanently.

A focused temporary regression reproduced this: after one 2-byte send occupied a zero-capacity queue, a second 2-byte send was polled to `Pending` and dropped; the counter remained 4 instead of returning to 2. Current CKB does not expose a proven attacker-controlled cancellation primitive for this exact future, so this is Low.

Recommendation: create an RAII reservation token immediately after reservation and move it with the payload through every queue so cancellation and queue teardown release the charge automatically.

### Validation

- `cargo test -p tentacle pending_data_tests --lib`: 7 passed
- `cargo test -p tentacle --test test_send_buffer_accounting`: 2 passed
- Focused cancellation regression: finding reproduced; temporary test removed after evidence collection
- Broad workspace tests were not run; PR CI remains the regression authority

File: tentacle/src/service/control.rs [hunk 1]
```diff
@@ -131,11 +197,13 @@ impl ServiceControl {
...
-        self.send(ServiceTask::ProtocolMessage {
-            target,
-            proto_id,
-            data,
-        })
+        if self.closed.load(Ordering::SeqCst) {
+            return Err(SendErrorKind::BrokenPipe);
+        }
+        match protocol_message_task(&self.sessions, target, proto_id, data) {
+            Some(task) => self.send(task),
```
Comment (zhangsoledad) [inline tentacle/src/service/control.rs:204]: This introduces a new failure path when the service queue is full. `protocol_message_task()` has already set the session's `closed` flag after a budget breach, but `send(task)` can return `WouldBlock` and drop the task carrying `blocked`. A retry then gets `PendingDataReservation::Closed`, so it cannot recreate the notification.

I reproduced this by filling the task queue, exceeding the same session's byte budget, draining the queue, and retrying. The queued task had no `blocked` entry, and the retry produced no new task to notify the service. The base doesn't set `closed` before queue admission, so this is introduced by the new reservation flow, separate from the existing late-accounting issue this PR fixes.

Let's keep the pending close request in state that can wake the service independently of the data queue, and add a regression covering the full-queue failure followed by a retry.

**Review by zhangsoledad — COMMENTED** [review]
There's a new edge case in the pre-queue reservation path when the service queue is full. I reran the regression against `7b091584` and reproduced the lost notification. Details inline.
